core/src/main/scala/org/apache/spark/metrics/ExecutorMetricType.scala

```diff
@@ -19,18 +19,31 @@ package org.apache.spark.metrics
 import java.lang.management.{BufferPoolMXBean, ManagementFactory}
 import javax.management.ObjectName
 
+import scala.collection.mutable
+
+import org.apache.spark.executor.ProcfsBasedSystems
 import org.apache.spark.memory.MemoryManager
 
 /**
  * Executor metric types for executor-level metrics stored in ExecutorMetrics.
  */
 sealed trait ExecutorMetricType {
-  private[spark] def getMetricValue(memoryManager: MemoryManager): Long
-  private[spark] val name = getClass().getName().stripSuffix("$").split("""\.""").last
+  private[spark] def getMetricValue(memoryManager: MemoryManager): Long = 0
+  private[spark] def getMetricSet(memoryManager: MemoryManager): Array[Long] = {
+    new Array[Long](0)
+  }
+  private[spark] def names = Seq(getClass().getName().stripSuffix("$").split("""\.""").last)
```

Comment:
this is not a good default -- at least not on this class, where the interface allows returning multiple metrics.

I do think it makes sense to have a child trait here just for all the MetricTypes which only have one value, which does have this default.  Also that would be the place to have an abstract `getMetricValue()` (singular) method.  That would avoid so many `new Array[Long](1)` etc. etc.

Reply:
Thank you @squito for the comments. Better to finalize this before doing back and force.
So what you suggest is that having two trait one for MetricTypes with one value (Lets call it SingleValueExecutorMetric) and one for MetricType with multiple Values (Lets call it MultiValueExecutorMetric). This I think will force us to have two metricGetters in ExecutorMetricType object as well, since one will call the getMetricValue and the other will call getMetricSet (or getMetricValues). Just wondering if I understand your suggestion correctly.

Reply:
I think I'm suggesting something slightly different.  There would be two traits, but they'd be related.   `trait ExecutorMetricType` would allow multiple values, as you have it here.  Everything which needed to interact with the metrics would use that interface.  That would define a `getMetricsValues()` etc. (plural).

Then you'd have another `trait SingleValueMetricType` (or something) which would have abstract methods like `getMetricValue()` singular, and it would provide a definition for the plural methods like

```scala
def getMetricValues(): Array[Long] = {
  val metrics = new Array[Long](1)
  metrics(0) = getMetricValue()
  metrics
}
```

in java terms, you could think of `ExecutorMetricType` as an interface and `SingleValueMetricType` as an abstract class.  `SingleValueMetricType` is just a convenience for some implementations to take advantage of to avoid repeating code.